Add EIP-6066: Signature Validation Method for NFTs (#6066)
ethereum/EIPs: EIPS/eip-6066.md

## Specification

_The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119._

```
pragma solidity ^0.8.0;

interface IEIP6066 {
    /**
     * @dev Should return whether the signature provided is valid for the provided tokenId and hash
     * @param tokenId   Token ID of the signing NFT
     * @param hash      Hash of the data to be signed
     *
     * MUST return the bytes4 magic value 0x1e6395e6 when function passes.
     * MUST NOT modify state (using STATICCALL for solc < 0.5, view modifier for solc > 0.5)
     * MUST allow external calls
     *
     */
    function isValidSignature(
        uint256 tokenId,

>         bytes32 hash

Reviewer: I am in full support for this EIP which extends EIP-1271, this is well needed! E.g. This EIP will be useful for [EIP-5453](https://eips.ethereum.org/EIPS/eip-5453) endorsement which. Feel free to mention use cases like EIP-5453 in your Motivation to add the weights of necessity of this EIP.

Also, I like to propose considering adding a extension field for future-proof (https://eips.ethereum.org/EIPS/eip-5750)
```suggestion
        bytes32 hash,
        bytes calldata extraData,
```

For example, this will allow future extending EIPs, e.g.
1.  add "only consider valid when user's EIP-1155 balance of tokenId > miniumBalance
2. only consider valid if user also hold slots of `slotId` 

The EIP-5750 documented the rationale why adding an extension is beneficial. Had EIP-1271 has this extending field, they will not have the disadvantage that this EIP intends to solve and improve.

In short, this is EIP a great idea.

Author: Thank you for the support and feedback, I have incorporated EIP-5750 into this EIP.

---
eip: 6066
title: Signature Validation Method for NFTs
description: A way to verify signatures when the signing entity is an EIP-721 or EIP-1155 NFT
author: Jack Boyuan Xu (@boyuanx)
discussions-to: https://ethereum-magicians.org/t/eip-6066-signature-validation-method-for-nfts/
status: Draft
type: Standards Track
category: ERC
created: 2022-11-29

> requires: 721, 1155, 5750

Reviewer: ```suggestion
requires: 721, 1155, 1271, 5750
```

---

## Abstract

While **E**xternally **O**wned **A**ccounts can validate signed messages with `ecrecover()` and smart contracts can validate signatures using specifications outlined in [EIP-1271](./eip-1271.md), currently there is no standard method to create or validate signatures made by NFTs. We propose a standard way for anyone to validate whether a signature made by an NFT is valid. This is possible via a modified signature validation function originally found in [EIP-1271](./eip-1271.md): `isValidSignature(tokenId, hash, data)`.

## Motivation

> With billions of ETH in trading volume, the **N**on-**F**ungible **T**oken standard has exploded into tremendous popularity in recent years. Despite the far-reaching implications of having unique tokenized items on-chain, NFTs have mainly been used to represent artwork in the form of avatars or profile pictures. While this is certainly not a trivial use case for the [EIP-721](./eip-721.md) & [EIP-1155](./eip-1155.md) token standards, we reckon more can be done to aid the community in discovering alternative uses for NFTs.

Reviewer 2: This section doesn't convince me, a skeptical reader, that this EIP is necessary. What use cases require a signature from the NFT itself, and not its current holder? What makes this proposal different from other signature schemes?

Author: This is most useful when the signing entity is a certain position within an organization or, broadly speaking, any identity construct that is not be tied to any singular address over time. It allows signatures to act as a stamp of approval representing the consent of an organization instead of any individual person. Currently, all consent represented by signatures is strictly tied to a single address that's non-transferrable. This EIP adds a layer of abstraction and provides flexibility when consent transfer is needed.

## Specification

> _The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119._

Reviewer: Update the latest template that superceed RFC 2119 with 8174 and please update keywords to "uppercase" below whenever you mean it
```suggestion
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “NOT RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119 and RFC 8174
```

Author: Thanks for the suggestions, I’ll get them revised :)

```
pragma solidity ^0.8.0;

interface IEIP6066 {
    /**
     * @dev Should return whether the signature provided is valid for the provided tokenId and hash
     * @param tokenId   Token ID of the signing NFT
     * @param hash      Hash of the data to be signed
     * @param data      Optional arbitrary data that may aid verification
     *
     * MUST return the bytes4 magic value 0xdff13226 when function passes.
     * MUST NOT modify state (using STATICCALL for solc < 0.5, view modifier for solc > 0.5)
     * MUST allow external calls
     *
     */
    function isValidSignature(
        uint256 tokenId,
        bytes32 hash,
        bytes calldata data
    ) external view returns (bytes4 magicValue);
}
```

`isValidSignature` can call arbitrary methods to validate a given signature.

> This function should be implemented by [EIP-721](./eip-721.md) or [EIP-1155](./eip-1155.md) compliant contracts that desire to enable its token holders to sign messages using their NFTs. Applications wanting to support contract signatures should call this method if the signer is an NFT.

Reviewer: I think this EIP can be generalized to apply to any contract with some sort of identifier inside. ERC721 is one of such contracts. Hence the suggestion below:
```suggestion
This function MAY be implemented by [EIP-721](./eip-721.md) or [EIP-1155](./eip-1155.md) compliant contracts that desire to enable its token holders to sign messages using their NFTs. Compliant Callers wanting to support contract signatures MUST call this method if the signer is holder of an NFT or ERC1155 token.
```

Author: Yeah, this can be generalized to any future NFT standard that involves a `tokenId`.

> 

Reviewer: Optional ideas for backward compatibility with ERC1271

```suggestion
3. When a Compliant Contract also implements ERC1271 signature, the signing algorithm MUST be the identical signing algorithm with its choice of ERC1271.
```

Author: Mmm, I'm not so sure about this one. There isn't an explicit "signing algorithm" for this EIP, as consent is provided by the fact that the owner of the NFT created, signed, and broadcasted the transaction that would flip the boolean flag given a hash.

Reviewer: Sure. I am open to not include this particular idea.

## Rationale

> One of the alternative use cases for NFTs is using them to represent offices in an organization. In this case, tying signatures to transferrable NFTs instead of EOAs or smart contracts becomes crucial. Suppose there exists a DAO that utilizes NFTs as badges that represent certain administrative offices (i.e., CEO, COO, CFO, etc.) with a quarterly democratic election that potentially replaces those who currently occupy said offices. If the sitting COO has previously signed agreements or authorized certain actions, their past signatures would stay with the EOA who used to be the COO instead of the COO's office itself once they are replaced with another EOA as the new COO-elect. Although a multisig wallet for the entire DAO is one way to mitigate this problem, often it is helpful to generate signatures on a more intricate level so detailed separation of responsibilities are established and maintained. It is also feasible to appoint a smart contract instead of an EOA as the COO, but the complexities this solution brings are unnecessary. If a DAO uses ENS to establish their organizational hierarchy, this proposal would allow wrapped ENS subdomains (which are NFTs) to generate signatures.

Reviewer 2: This paragraph reads more like motivation (and not rationale.) The rationale should describe individual decisions made within the EIP itself (the next paragraph is a perfect example), while the Motivation justifies the existence of the EIP in its entirety.

Author: Thanks, I'll work on this more.
